release/air_tests/air_benchmarks/workloads/torch_benchmark.py

```diff
@@ -0,0 +1,298 @@
+    )
+    result = trainer.fit()
+    print(f"Last result: {result.metrics}")
+
+
+def train_torch_vanilla_worker(
+    *, config: dict, rank: int, world_size: int, master_addr: str, master_port: int
+):
+    # This function is kicked off by the main() function and runs the vanilla
+    # training script on a single worker.
+    os.environ["MASTER_ADDR"] = master_addr
+    os.environ["MASTER_PORT"] = str(master_port)
+    distributed.init_process_group("gloo", rank=rank, world_size=world_size)
+
+    train_func(use_ray=False, config=config)
+
+    distributed.destroy_process_group()
+
+
+def train_torch_vanilla(*, config: dict, num_workers: int = 4, use_gpu: bool = False):
+    # This function is kicked off by the main() function and subsequently kicks
+    # off tasks that run train_torch_vanilla_worker() on the worker nodes.
+    import ray
+    from benchmark_util import upload_file_to_all_nodes, run_command_on_all_nodes
+
+    path = os.path.abspath(__file__)
+    upload_file_to_all_nodes(path)
+    master_addr = ray.util.get_node_ip_address()
+    master_port = 12355  # hardcoded
+
+    # Each worker needs to know which rank it is. To pass this, we construct a
+    # map of IPs to ranks, which is passed as a string to all workers. The workers
+    # then parse this string to determine their local rank.
+    node_ip_to_ranks = {
+        ip: rank
+        for rank, ip in enumerate(
+            [
+                n["NodeManagerAddress"]
+                for n in ray.nodes()
+                if n["Alive"] and n["NodeManagerAddress"] != master_addr
+            ],
+            start=1,
+        )
+    }
+    node_ip_to_ranks[master_addr] = 0
+    node_to_rank_str = ",".join(
+        {f"{ip}:{rank}" for ip, rank in node_ip_to_ranks.items()}
+    )
+
+    num_epochs = config["epochs"]
+
+    cmd = [
+        "python",
+        path,
+        "--vanilla-worker",
+        "--num-epochs",
+        str(num_epochs),
+        "--node-to-rank-str",
+        node_to_rank_str,
+        "--num-workers",
+        str(num_workers),
+        "--master-addr",
+        master_addr,
+        "--master-port",
+        str(master_port),
+    ]
+
+    if use_gpu:
+        cmd += ["--use-gpu"]
+
+    run_command_on_all_nodes(cmd)
+
+
+@click.command()
+@click.option("--num-workers", type=int, default=4)
+@click.option("--num-epochs", type=int, default=4)
+@click.option("--use-gpu", is_flag=True, default=False)
+@click.option("--vanilla-worker", is_flag=True, default=False)
+@click.option("--node-to-rank-str", type=str, default="")
+@click.option("--master-addr", type=str, default="")
+@click.option("--master-port", type=int, default=0)
+def main(
+    num_workers: int = 4,
+    num_epochs: int = 4,
+    use_gpu: bool = False,
+    vanilla_worker: bool = False,
+    node_to_rank_str: str = "",
+    master_addr: str = "",
+    master_port: int = 0,
+):
+    config = {"lr": 1e-3, "batch_size": 64, "epochs": num_epochs}
+
+    if vanilla_worker:
```

Comment:
Minor point, but for the vanilla case would it be better to have the driver code and the worker code in separate functions for better readability?

Reply:
It seems like currently `main` calls `train_torch_vanilla` which calls `main` again for each worker.

Reply:
It calls main, but in a new process. Putting these into separate files runs into import errors, but I'd like to keep the training loops in the same file so we can be sure exactly the same training loop is executed

Reply:
Got it- I understand it's in a new process. It can be in the same file, but do you think it makes sense to split in separate functions?

Reply:
Not sure what you mean - that's what `train_torch_vanilla` (driver) and `train_torch_vanilla_worker` (worker) are for, right? `main()` is just the main entrypoint

Reply:
Ah I see, you mean splitting it into a click group and subcommands? Sure I can do that